### PR TITLE
Initial Output Manager Pool Population

### LINKED ISSUE #219: Remove print_pools function from Output_Manager prior to merging PR #218
Created a temporary `print_pools` method at the end of the OutputManager class which is called at the end of a simulation to show OutputManager is indeed collecting data being sent to it and will show some examples of how that raw data looks in the pools (and after prefixes/suffixes have been appended).

This is serving as a reminder to delete this method from the OutputManager class prior to merging the PR it is created in. There is a TODO in the code as well right above the `print_pools` method. There is a call to `print_pools` in the SimulationEngine.simulate method which will need to be deleted prior to merging as well. I have added a TODO above that call as a reminder.

We are trying to reduce the number of overall print statements in the model so this is not a normal addition and is purely for the aid of the PR reviewers as an easy view of the OutputManager pools output.

### PR DESCRIPTION
**Pull request by ew3361zh**
- Added a first round of variables, warnings, and errors to Output Manager pools
- Primarily based on searches for print statements in the code with the idea that the Output Manager will eventually replace the need for those print statements.
- Created a (temporary) `print_pools` method in the OutputManager class which is called at the end of a simulation to show OutputManager is indeed collecting data being sent to it and will show some examples of how that raw data looks in the pools (and after prefixes/suffixes have been appended). This method will almost definitely not stick around but I found it useful in seeing the Output Manager's pool collection system in action.

### CODE REVIEW

**Review by PooyaHekmati — CHANGES_REQUESTED** [review]
As we have discussed in our meetings, these need to be addressed:

1. Import statements should be sorted in a-z order, so change the following (and similar cases)
`from RUFAS.util import Utility`
`from RUFAS.output_manager import OutputManager`
to
`from RUFAS.output_manager import OutputManager`
`from RUFAS.util import Utility`
3. Avoid using hard-coded values (see the example below)
4. Avoid using long lines (a line shouldn't be longer than 99 chars, preferably less than 79 chars)
5. Avoid adding new print statements
6. Signals on the PR are failing (linting failed), which needs to be fixed
7. info_map is meant to capture environment variables to have reproducible results, warnings, and errors. Just adding caller_class and caller_function is the bare minimum and is NOT a good practice. 
8. When adding errors and warnings, remove the corresponding print statements.
9. A PR should move the code base toward being production-ready. If a temporary code snippet is added, a proper plan should be devised for its removal. This plan has to be documented using git issues (needs to be mentioned as a TODO in the code) and communicated in the PR description.
10. Practical examples of output manager usage are available in the output_manager.py and its corresponding test cases; here is another example:
`from RUFAS.output_manager import OutputManager`
`om = OutputManager()`
`class foo (object):`
`    def bar(self, arg1, arg2):`
`        info_map={'caller_class': self.__class__.__name__,`
`                 'caller_function': self.bar.__name__,`
`                 'arg1':arg1,`
`                 'arg2':arg2,`
`                 }`
`        om.add_log('example_log',`
`                    'This is an example to demonstrate how'`
`                    +' a long string should be created and sent'`
`                    +' to output manager. Notice that + signs'`
`                    +' are at the beginning of each new line'`
`                    +' rather at the end of last line. This'`
`                    +' enhances readability and reduces confusion.',`
`                    info_map)`         
`        variable_x = some value`
`        variable_y = something else`
`        some logic that changes x and y`
`        om.add_variable('variable_x',variable_x,info_map)`
`        some logic that leads to a situation that needs user's attention`
`        info_map['variable_x']=variable_x`
`        info_map['variable_y']=variable_y`
`        om.add_warning('exmaple_warning',`
`                       'The values of x and y are not making sense!',`
`                       info_map)`
`        #now that we have added x and y to the info_map, moving forward,`
`        #they will always be logged. This may not be desired as it can lead`
`        #to a situation where there are so many keys in the info_map which`
`        #makes debugging difficult. To avoid this, instead of adding x and y`
`        #to info map, we could first call add_variable for them and then`
`        #add the warning without adding them to the info map. If this path`
`        #is used, then the warning message should clearly point out that the`
`        #last values of x and y led to this warning. Alternatively, we can`
`        #create a new info_map. Note that the name of this dict can be anything,`
`        #doesn't need to be info_map. We can also use the pop() method to remove`
`        #a key from a dict e.g., info_map. The samething applies to errors as well.`

**Comment by ew3361zh** [comment]
> As we have discussed in our meetings, these need to be addressed:
> 
> 1. Import statements should be sorted in a-z order, so change the following (and similar cases)
>    `from RUFAS.util import Utility`
>    `from RUFAS.output_manager import OutputManager`
>    to
>    `from RUFAS.output_manager import OutputManager`
>    `from RUFAS.util import Utility`
> 2. Avoid using hard-coded values (see the example below)
> 3. Avoid using long lines (a line shouldn't be longer than 99 chars, preferably less than 79 chars)
> 5. Signals on the PR are failing (linting failed), which needs to be fixed
> 7. When adding errors and warnings, remove the corresponding print statements.

I will fix these today (and going forward).

> 4. Avoid adding new print statements
> 8. A PR should move the code base toward being production-ready. If a temporary code snippet is added, a proper plan should be devised for its removal. This plan has to be documented using git issues (needs to be mentioned as a TODO in the code) and communicated in the PR description.

I believe the only print statements added were for the temporary print_pools function which was added for the benefit of the reviewers to see what the gathered data would look like and have the option to provide feedback on whether that was capturing the data as they'd hoped. It's noted in the PR these will be removed prior to merging but I will add a git issue and in-code TODO statements as well.

> 6. info_map is meant to capture environment variables to have reproducible results, warnings, and errors. Just adding caller_class and caller_function is the bare minimum and is NOT a good practice.

Ok, this was a misunderstanding of the purpose of the info_map on my part. Perhaps we can discuss further in our check-in as I have some follow-up questions to what is explained in the example (9) below.

> 9. Practical examples of output manager usage are available in the output_manager.py and its corresponding test cases; here is another example:
>    `from RUFAS.output_manager import OutputManager`
>    `om = OutputManager()`
>    `class foo (object):`
>    `    def bar(self, arg1, arg2):`
>    `        info_map={'caller_class': self.__class__.__name__,`
>    `                 'caller_function': self.bar.__name__,`
>    `                 'arg1':arg1,`
>    `                 'arg2':arg2,`
>    `                 }`
>    `        om.add_log('example_log',`
>    `                    'This is an example to demonstrate how'`
>    `                    +' a long string should be created and sent'`
>    `                    +' to output manager. Notice that + signs'`
>    `                    +' are at the beginning of each new line'`
>    `                    +' rather at the end of last line. This'`
>    `                    +' enhances readability and reduces confusion.',`
>    `                    info_map)`
>    `        variable_x = some value`
>    `        variable_y = something else`
>    `        some logic that changes x and y`
>    `        om.add_variable('variable_x',variable_x,info_map)`
>    `        some logic that leads to a situation that needs user's attention`
>    `        info_map['variable_x']=variable_x`
>    `        info_map['variable_y']=variable_y`
>    `        om.add_warning('exmaple_warning',`
>    `                       'The values of x and y are not making sense!',`
>    `                       info_map)`
>    `        #now that we have added x and y to the info_map, moving forward,`
>    `        #they will always be logged. This may not be desired as it can lead`
>    `        #to a situation where there are so many keys in the info_map which`
>    `        #makes debugging difficult. To avoid this, instead of adding x and y`
>    `        #to info map, we could first call add_variable for them and then`
>    `        #add the warning without adding them to the info map. If this path`
>    `        #is used, then the warning message should clearly point out that the`
>    `        #last values of x and y led to this warning. Alternatively, we can`
>    `        #create a new info_map. Note that the name of this dict can be anything,`
>    `        #doesn't need to be info_map. We can also use the pop() method to remove`
>    `        #a key from a dict e.g., info_map. The samething applies to errors as well.`

**Comment by ew3361zh** [comment]
@PooyaHekmati if you could review again, I believe I've addressed the changes requested

**Review by morrowcj — COMMENTED** [review]
This looks good. I looked through the changes to see how you were using the output manager in practice. I won't formally approve, since I think it is probably important for someone from the animal management team to do so.

File: RUFAS/output_manager.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 from typing import Any, Dict, List, Union
 import time
+from pprint import pprint
```
Comment (PooyaHekmati) [inline RUFAS/output_manager.py]: Is this still needed?

File: RUFAS/routines/animal/animal_management.py [hunk 2]
```diff
@@ -218,7 +233,10 @@ def init_pens(self, all_pens_data, herd_data):
             self.all_pens.append(pen_2)
             self.all_pens.append(pen_3)
         elif (len(self.all_pens) == 2) and (herd_num > 0):
-            print('Warning: herd_num > 0, but pen_num = 2. Initilizing 1 default pen.')
+            om.add_warning('invalid_pen_num_warning', 
+                            'Warning: herd_num > 0, but pen_num = 2.' 
+                            + ' Initilizing 1 default pen.',
+                             {'caller_class': 'AnimalManagement', 'caller_function': 'init_pens'})
```
Comment (PooyaHekmati) [inline RUFAS/routines/animal/animal_management.py]: Avoid using hardcoded values

File: RUFAS/simulation_engine.py [hunk 3]
```diff
@@ -26,22 +29,30 @@ def __init__(self, input_file_path: Path) -> None:
...
+                    'caller_function': self.simulate.__name__,}
         t_start_sim = timer.time()
+        
         self._run_simulation_main_loop()
         self.output.finalize(self.state, self.weather, self.time)
         t_end_sim = timer.time()
 
         print("Simulation Successful")
         print(f"Total Simulation Time: {t_end_sim - t_start_sim} seconds")
+        total_simulation_time = t_end_sim - t_start_sim
```
Comment (PooyaHekmati) [inline RUFAS/simulation_engine.py]:
```suggestion
        total_simulation_time = t_end_sim - t_start_sim
        print(f"Total Simulation Time: {total_simulation_time } seconds")
```

File: RUFAS/simulation_engine.py [hunk 4]
```diff
@@ -26,22 +29,30 @@ def __init__(self, input_file_path: Path) -> None:
...
+        om.add_variable('total_simulation_time', total_simulation_time, info_map)
```
Comment (PooyaHekmati) [inline RUFAS/simulation_engine.py]: I think this is more of a log than a variable

File: RUFAS/simulation_engine.py [hunk 5]
```diff
@@ -26,22 +29,30 @@ def __init__(self, input_file_path: Path) -> None:
...
+
 
         t_start_graphics = timer.time()
         sys.stdout.write('Producing Graphics\n')
         self.output.produce_graphics()
         t_end_graphics = timer.time()
 
         graphics_prod_time = t_end_graphics - t_start_graphics
+        om.add_variable('graphics_prod_time', graphics_prod_time, info_map)
```
Comment (PooyaHekmati) [inline RUFAS/simulation_engine.py]: Same, can this be a log?

File: RUFAS/simulation_engine.py [hunk 6]
```diff
@@ -26,22 +29,30 @@ def __init__(self, input_file_path: Path) -> None:
...
         total_runtime = (t_end_sim-t_start_sim) + \
             (t_end_graphics-t_start_graphics)
+        om.add_variable('total_runtime', total_runtime, info_map)
```
Comment (PooyaHekmati) [inline RUFAS/simulation_engine.py]: log?

File: RUFAS/simulation_engine.py [hunk 7]
```diff
@@ -141,8 +152,12 @@ def _initialize_simulation(self, file_path: Path) -> None:
             InvalidJSONFile: If the json file at the given path does not conform 
             with the format required
         """
-
+        info_map = {'caller_class': self.__class__.__name__, 
+                    'caller_function': self._initialize_simulation.__name__,
+                    'file_path': file_path,}
         print(f"Initializing simulation environment from {file_path}")
+        om.add_variable('simulation_initialization_file_path', file_path, info_map)
```
Comment (PooyaHekmati) [inline RUFAS/simulation_engine.py:158]: log?

Comment (PooyaHekmati) [reply RUFAS/simulation_engine.py:158]: @JoeWaddell What are your thoughts here? When running bulk simulations, are these variables or logs?

File: RUFAS/user_prompt.py [hunk 8]
```diff
@@ -9,8 +9,10 @@
 from pathlib import Path
 
 from RUFAS import util, errors
+from RUFAS.output_manager import OutputManager
 import fileReader
 
+om = OutputManager()
```
Comment (PooyaHekmati) [inline RUFAS/user_prompt.py]: I don't see any usage of output manager here, do we need this?

File: RUFAS/util.py [hunk 9]
```diff
@@ -11,7 +11,9 @@
 from typing import Callable, List, Tuple
 
 from RUFAS import errors
+from RUFAS.output_manager import OutputManager
 
+om = OutputManager()
```
Comment (PooyaHekmati) [inline RUFAS/util.py]: same, no usage

**Review by PooyaHekmati — COMMENTED** [review]
LGTM from a code quality perspective. I have some comments and questions, waiting for the author's response.

File: RUFAS/simulation_engine.py (hunk 7 above)
Comment (JoeWaddell) [reply RUFAS/simulation_engine.py:158]: Great question: I could be convinced either way, but I think in my particular case saving it as a variable would be useful in my workflow. 
When I run simulations I use a key that associates the runs with a given suite of input parameters. Having the `file_path` saved as a variable could be useful to print in an output, just to ensure that things aren't getting mixed up when moving/copying files. 
I could also see it being useful to keep this as a variable for users who might want to run multiple simulations testing different scenarios: could help avoid problems from sloppy file management after the fact.

File: RUFAS/user_prompt.py (hunk 8 above)
Comment (JoeWaddell) [reply RUFAS/user_prompt.py]: I was going to ask about this section as well (was reviewing it as these comments came in). Further down in this file, the `info_map` is set, but not used. I assumed it is a placeholder for adding logs/warnings, or somehow being used in a way that eluded me.

**Review by JoeWaddell — COMMENTED** [review]
Nice work! I've folded the couple comments I had into replies to Pooya's, as this all looks good otherwise. Seems like a nice demonstration of how this will work moving forward.

File: RUFAS/simulation_engine.py (hunk 7 above)
Comment (PooyaHekmati) [reply RUFAS/simulation_engine.py:158]: Thanks @JoeWaddell . What about 'total_runtime', 'total_simulation_time', and 'graphics_prod_time' ? See the comments above. Do you see them as logs or variables?

Comment (JoeWaddell) [reply RUFAS/simulation_engine.py:158]: @PooyaHekmati : for my own purposes, I think I'd treat those as logs. 
Likewise, for a typical user, I can't think of a good reason to treat them as variables. 
For advanced users who want to explore the runtimes for different farm sizes or longer time periods, I imagine checking the logs should be sufficient in those cases.

File: RUFAS/output_manager.py (hunk 1 above)
Comment (ew3361zh) [reply RUFAS/output_manager.py]: no longer needed, deleted.

File: RUFAS/routines/animal/animal_management.py (hunk 2 above)
Comment (ew3361zh) [reply RUFAS/routines/animal/animal_management.py]: switched to info_map variable to match other usages

File: RUFAS/simulation_engine.py (hunk 3 above)
Comment (ew3361zh) [reply RUFAS/simulation_engine.py]: addressed this in the commit for all the fixes suggested.

File: RUFAS/simulation_engine.py (hunk 4 above)
Comment (ew3361zh) [reply RUFAS/simulation_engine.py]: agreed, switched method to add_log

File: RUFAS/simulation_engine.py (hunk 5 above)
Comment (ew3361zh) [reply RUFAS/simulation_engine.py]: agreed, switched method to add_log

File: RUFAS/simulation_engine.py (hunk 6 above)
Comment (ew3361zh) [reply RUFAS/simulation_engine.py]: switched method to add_log

File: RUFAS/simulation_engine.py (hunk 7 above)
Comment (ew3361zh) [reply RUFAS/simulation_engine.py:158]: With the calculated times being added as logs are you both okay with eliminating the call to _show_final_messages() at the end of the simulate method?

File: RUFAS/user_prompt.py (hunk 8 above)
Comment (ew3361zh) [reply RUFAS/user_prompt.py]: Thank you, Joe and Pooya! These were leftovers from my initial run through when I was adding the raised errors to the errors pool but after discussions with Pooya it was decided that this was not a useful type of error to be included.

File: RUFAS/util.py (hunk 9 above)
Comment (ew3361zh) [reply RUFAS/util.py]: It has been removed.

File: RUFAS/simulation_engine.py (hunk 7 above)
Comment (PooyaHekmati) [reply RUFAS/simulation_engine.py:158]: I think it is better to leave it as is for now. We will come back and revisit these kinds of issues later.

**Review by PooyaHekmati — APPROVED** [review]

**Review by JoeWaddell — APPROVED** [review]

**Pull request merged**